### DISCUSSION TITLE
Beta versions

### DISCUSSION THREAD

**Discussion by infamous-panda** [terjeio/ioSender#155] [General]
I am trying to root out some speed issues between my new setup using the later GRBL-HAL build + iosender 2.0.34 and a machine built a year and half ago using gcode sender beta 7 and beta 8. I can't seem to find these versions online anymore. Anyway to put a link up for them?

**Comment by terjeio** [comment]
I can if I have a backups with those builds on them. I'll check later.

**Reply by terjeio** [reply]
You can find them [here](http://www.io-engineering.com/downloads), but not for very long.

**Comment by infamous-panda** [comment]
Terje,

I am still testing a few other variables but one thing I am finding somewhat consistently is that was gettting much better performance from the older gcode sender betas.

For example I have a test file which I time to completion which took gcode sender Beta7 about 1:22 to complete at an average rate of about 54,000 mm/m. IO sender 2.0.34 takes about +10 minutes at an average rate of 6,000 mm/m. 

I can test other variables and provide more detail but maybe there something obvious I may be missing. This is with the same teensy controller with the motors disconnected.

**Reply by terjeio** [reply]
I get the same feed rates with 2.0.36 and beta 7, processor load is about the same for both versions at around 36%. When I run off the SD-card feed rates are again the same but processor load is significantly lower at less than 2%.

Can you check processor load?
And is running from the SD card slow with 2.0.34 too?

Is the test file the same as we used before? I am testing with _CONTROLLER.SPEED.TEST.NOS.txt_

**Comment by infamous-panda** [comment]
Hi Terjeio,

I did some more research on this. I had forgotten that I started an email thread with Phil touching on some of the same items around this time last year when I was closing out my production laser. I took the board board out of our production laser and put it through the same testing as the two other boards I have. Here are the main points that I found. 

1. Our production Laser has been running on firmware I compiled from the old GRBL-HAL Github page under releases which apparently was different from what was posted under master in November 2020. Looks like Build 20200503 from the reporting. No ethernet or SDcard enabled.
2. Our production Laser uses an old windows laptop sending via Gcode Sender 8.
3. We regularly run files similar to the controller speed tests and sustain 50,000 mm/m
4. Completion time for our controller speed test file is 1:22 at 50,000 mm/m max speed and 800 mm/m accelleration
5. If we disconnect the motors and increase speed and acceleration to 90,000, we can sustain a stable 90,000 mm/m and finish the file in 50 seconds. **This is streaming from USB!** and basically equal to the speed we get when streaming from the SD card. 
6. The best we have gotten with any combination of the newer IO Sender and the current build of GRBL-HAL is an erratic 12-38K mm/m with an average finish time of 10-11 minutes. 

I laid out the test parameters in the spreadsheet below

https://docs.google.com/spreadsheets/d/1JXFqQkkx0XezlIa1N2vBQ7xxXwEVvD3wcgtJPfnJQEA/edit#gid=1360807085

Files I used here.

https://drive.google.com/drive/folders/16mDC0mzsZiBIr2t34PtIerFRpnp5mTan?usp=sharing

Hopefully you have some time to look at this as I am sure there is something interesting going on. We have been trying to put together some additional laser test beds to experiment with the newer firmware and senders but they are surprisingly an order of magnitude slower than what we have been running all year.

**Reply by terjeio** [reply]
I am not able to replicate this, can you post the controller settings? _Use Help > About > To clipboard_ and paste them in a comment.

What makes a huge difference in speed is if _$32 Laser mode_ is enabled or not. If not enabled the planner buffer is emptied before the speed is changed. This results in the feed rate not beeing reported for some reason, I guess because the feed rate is actually 0 when the report is generated.
BTW with laser mode enabled and switching the laser on with M4 S commands can be removed when used to switch the laser on/off as done in the test file - G1/G0 does the same but faster.

**Reply by infamous-panda** [reply]
> I am not able to replicate this, can you post the controller settings? _Use Help > About > To clipboard_ and paste them in a comment.
> 
> What makes a huge difference in speed is if _$32 Laser mode_ is enabled or not. If not enabled the planner buffer is emptied before the speed is changed. This results in the feed rate not beeing reported for some reason, I guess because the feed rate is actually 0 when the report is generated. BTW with laser mode enabled and switching the laser on with M4 S commands can be removed when used to switch the laser on/off as done in the test file - G1/G0 does the same but faster.

Settings for the board with current firmware

%
; grblHAL
; 1.1f.20210715
; [OPT:VNMSL,999,1024,4,0]
; [NEWOPT:ENUMS,RT+,ES,TC]
; [FIRMWARE:grblHAL]
; [NVS STORAGE:*FLASH]
; [DRIVER:iMXRT1062]
; [DRIVER VERSION:210725]
; [DRIVER OPTIONS:USB.2]
; [BOARD:T41U5XBB]
; [AUX IO:4,3,0,0]
;
$N0=
$N1=
; 0 - Step pulse time
$0=3.0
; 1 - Step idle delay
$1=25
; 2 - Step pulse invert
$2=0
; 3 - Step direction invert
$3=0
; 4 - Invert step enable pin(s)
$4=11
; 5 - Invert limit pins
$5=0
; 6 - Invert probe pin
$6=0
; 7 - Disable spindle with zero speed
$7=0
; 10 - Status report options
$10=511
; 11 - Junction deviation
$11=0.010
; 12 - Arc tolerance
$12=0.002
; 13 - Report in inches
$13=0
; 14 - Invert control pins
$14=0
; 15 - Invert coolant pins
$15=0
; 16 - Invert spindle signals
$16=0
; 17 - Pullup disable control pins
$17=0
; 18 - Pullup disable limit pins
$18=0
; 19 - Pullup disable probe pin
$19=0
; 20 - Soft limits enable
$20=0
; 21 - Hard limits enable
$21=0
; 22 - Homing cycle
$22=0
; 23 - Homing direction invert
$23=0
; 24 - Homing locate feed rate
$24=25.0
; 25 - Homing search seek rate
$25=500.0
; 26 - Homing switch debounce delay
$26=250
; 27 - Homing switch pull-off distance
$27=1.000
; 28 - G73 Retract distance
$28=0.100
; 29 - Pulse delay
$29=0.0
; 30 - Maximum spindle speed
$30=1000.000
; 31 - Minimum spindle speed
$31=0.000
; 32 - Mode of operation
$32=1
; 33 - Spindle PWM frequency
$33=5000
; 34 - Spindle PWM off value
$34=0.0
; 35 - Spindle PWM min value
$35=0.0
; 36 - Spindle PWM max value
$36=100.0
; 37 - Steppers deenergize
$37=0
; 39 - Enable legacy RT commands
$39=1
; 40 - Limit jog commands
$40=0
; 43 - Homing passes
$43=1
; 44 - Axes homing, first pass
$44=4
; 45 - Axes homing, second pass
$45=3
; 46 - Axes homing, third pass
$46=0
; 47 - Axes homing, fourth pass
$47=0
; 62 - Sleep enable
$62=0
; 63 - Feed hold actions
$63=3
; 64 - Force init alarm
$64=0
; 65 - Probing feed override
$65=0
; 80 - Spindle P-gain
$80=1.000
; 81 - Spindle I-gain
$81=0.010
; 82 - Spindle D-gain
$82=0.000
; 84 - Spindle PID max error
$84=0.000
; 85 - Spindle PID max I error
$85=10.000
; 90 - Spindle sync P-gain
$90=0.000
; 91 - Spindle sync I-gain
$91=0.000
; 92 - Spindle sync D-gain
$92=0.000
; 95 - Spindle sync PID max I error
$95=0.000
; 100 - X-axis travel resolution
$100=80.000
; 101 - Y-axis travel resolution
$101=80.000
; 102 - Z-axis travel resolution
$102=250.000
; 103 - A-axis travel resolution
$103=80.000
; 110 - X-axis maximum rate
$110=90000.000
; 111 - Y-axis maximum rate
$111=10000.000
; 112 - Z-axis maximum rate
$112=500.000
; 113 - A-axis maximum rate
$113=90000.000
; 120 - X-axis acceleration
$120=90000.000
; 121 - Y-axis acceleration
$121=1000.000
; 122 - Z-axis acceleration
$122=10.000
; 123 - A-axis acceleration
$123=90000.000
; 130 - X-axis maximum travel
$130=200.000
; 131 - Y-axis maximum travel
$131=200.000
; 132 - Z-axis maximum travel
$132=200.000
; 133 - A-axis maximum travel
$133=200.000
; 160 - X-axis backlash compensation
$160=0.000
; 161 - Y-axis backlash compensation
$161=0.000
; 162 - Z-axis backlash compensation
$162=0.000
; 163 - A-axis backlash compensation
$163=0.000
; 341 - Tool change mode
$341=0
; 342 - Tool change probing distance
$342=30.0
; 343 - Tool change locate feed rate
$343=25.0
; 344 - Tool change search seek rate
$344=200.0
; 345 - Tool change probe pull-off rate
$345=100.0
; 347 - Dual axis length fail
$347=5.0
; 348 - Dual axis length fail min
$348=2.500
; 349 - Dual axis length fail max
$349=25.000
; 370 - Invert I/O Port inputs
$370=0
; 372 - Invert I/O Port outputs
$372=0
%

Settings for the older firmware

; 
; 1.1f(Teensy 4.0).20200503
; [OPT:VNMSLW,1023,1024,4]
;
$N0=
$N1=
; 0 - Step pulse time
$0=3
; 1 - Step idle delay
$1=25
; 2 - Step pulse invert
$2=0
; 3 - Step direction invert
$3=0
; 4 - Invert step enable pin
$4=11
; 5 - Invert limit pins
$5=0
; 6 - Invert probe pin
$6=0
; 10 - Status report options
$10=9
; 11 - Junction deviation
$11=0.010
; 12 - Arc tolerance
$12=0.002
; 13 - Report in inches
$13=0
$14=0
$15=0
$16=0
$17=0
$18=0
$19=0
; 20 - Soft limits enable
$20=0
; 21 - Hard limits enable
$21=0
; 22 - Homing enable
$22=0
; 23 - Homing direction invert
$23=0
; 24 - Homing locate feed rate
$24=25.000
; 25 - Homing search seek rate
$25=500.000
; 26 - Homing switch debounce delay
$26=250
; 27 - Homing switch pull-off distance
$27=1.000
$28=0.100
$29=0
; 30 - Maximum spindle speed
$30=1000.
; 31 - Minimum spindle speed
$31=0.
; 32 - Laser mode
$32=1
$33=5000.000
$34=0.000
$35=0.000
$36=100.000
$37=0
$39=1
$40=0
$41=0
$42=2
$43=1
$44=4
$45=3
$46=0
$47=0
$56=5.000
$57=100.000
$58=-5.000
$59=500.000
$60=0
$61=0
$62=0
$63=3
$64=0
$65=0
; 100 - X-axis travel resolution
$100=80.000
; 101 - Y-axis travel resolution
$101=80.000
; 102 - Z-axis travel resolution
$102=250.000
; 103 - A-axis travel resolution
$103=80.000
; 110 - X-axis maximum rate
$110=90000.000
; 111 - Y-axis maximum rate
$111=10000.000
; 112 - Z-axis maximum rate
$112=500.000
; 113 - A-axis maximum rate
$113=90000.000
; 120 - X-axis acceleration
$120=90000.000
; 121 - Y-axis acceleration
$121=1000.000
; 122 - Z-axis acceleration
$122=10.000
; 123 - A-axis acceleration
$123=90000.000
; 130 - X-axis maximum travel
$130=200.000
; 131 - Y-axis maximum travel
$131=200.000
; 132 - Z-axis maximum travel
$132=200.000
; 133 - A-axis maximum travel
$133=200.000

**Reply by terjeio** [reply]
Thanks. Can you run some tests in Check mode? This basically gets us the time it takes to get the data across and update the UI.
I get just above 2 minutes, when disabling program list scrolling it drops down to just under one minute. FYI running Check mode from the SD card takes somewhere between 3 and 4 seconds. Note that check mode will fail with the old firmware due to a bug, perhaps with your new version too as it is from July 15th.

**Reply by infamous-panda** [reply]
> Thanks. Can you run some tests in Check mode? This basically gets us the time it takes to get the data across and update the UI. I get just above 2 minutes, when disabling program list scrolling it drops down to just under one minute. FYI running Check mode from the SD card takes somewhere between 3 and 4 seconds. Note that check mode will fail with the old firmware due to a bug, perhaps with your new version too as it is from July 15th.

Sure, How does one do test mode?

**Reply by terjeio** [reply]
![image](https://user-images.githubusercontent.com/20260062/141928596-48306b55-286b-4294-8cd2-06894559d97a.png)

**Reply by terjeio** [reply]
I have found a change in the code that slows down the max USB throughput. This is significantly slows down Check mode and will add some overhead in normal mode depending on the gcode.
Check mode time went from approx. 2 mins to 18 seconds for me when I reverted this change. I made some changes to the sender as well and managed to get it down to 12 seconds - but these messed up the program listing.

If you want to revert the change and test then replace [this code](https://github.com/grblHAL/iMXRT1062/blob/1426213fdd93411469d9e86d7b14af79d0a2d05c/grblHAL_Teensy4/src/driver.c#L2149-L2175) with:

```
static void execute_realtime (uint_fast16_t state)
{
    if(usb_serial_available())
        usb_execute_realtime(state);

#if ADD_MSEVENT
    if(ms_event) {

        ms_event = false;

  #if QEI_ENABLE
        if(qei.vel_timeout && !(--qei.vel_timeout)) {
            qei.encoder.velocity = abs(qei.count - qei.vel_count) * 1000 / (millis() - qei.vel_timestamp);
            qei.vel_timestamp = millis();
            qei.vel_timeout = QEI_VELOCITY_TIMEOUT;
            if((qei.encoder.event.position_changed = !qei.dbl_click_timeout || qei.encoder.velocity == 0))
                hal.encoder.on_event(&qei.encoder, qei.count);
            qei.vel_count = qei.count;
        }

        if(qei.dbl_click_timeout && !(--qei.dbl_click_timeout)) {
            qei.encoder.event.click = On;
            hal.encoder.on_event(&qei.encoder, qei.count);
        }
  #endif
    }
#endif // ADD_MSEVENT

#if ETHERNET_ENABLE
    grbl_enet_poll();
#endif
}
```

I'll try to make a load plugin later, having a large planner buffer combined with high speed movement is bound to hit the roof in terms of load at some point. It would be interesting to find out where the limit is so the controller can be tuned to stay within it.

For those interested: 
The file transferred is 3.413.756 bytes long, at 18 seconds this corresponds to a transfer rate of nearly 190 Kbytes/second. With a USB <> UART serial chip in between the PC and the MCU that operates at 115Kbaud it takes nearly 5 minutes to transfer the same amount of data...

**Reply by infamous-panda** [reply]
What a difference a couple lines make!! I added a couple of runs to the spreadsheet. The first test at 90K mm/m through USB matched what I saw with the old firmware and sender combo. I did a few just to see how much further I could push, and I ended having to use a faster PC to do so. The interface stutters more and more once I pass 125k to 200k. At 300k there is no updates or speed reporting sender but the file still completes in about 20 seconds. Seems like I could push further but 100k is probably a practical physical limit for now. Amazing! Thank you so much!

**Comment by infamous-panda** [comment]
I included a HEX file compiled from my old firmware in the Google drive folder if it helps. When I did this last year it was my first time doing anything like compiling firmware and I am sure I made a few mistakes on the sources. However it seems to have something that's letting stream very fast.

**Reply by terjeio** [reply]
I am not sure - I had some runs earlier where the job abrubtly terminated at around 50 seconds not even half finished, wonder what that was due too. And I guess I should try to come up with some method to measure the CPU load - perhaps by timing the idle loop?

Anyway - it is bedtime here so not now.

**Reply by infamous-panda** [reply]
> I am not sure - I had some runs earlier where the job abrubtly terminated at around 50 seconds not even half finished, wonder what that was due too. And I guess I should try to come up with some method to measure the CPU load - perhaps by timing the idle loop?
> 
> Anyway - it is bedtime here so not now.

Condition # 6 on my spreadsheet would not complete either and would stop at random points sometimes very early. Error:24 I believe. This is using a more current firmware and Gcode sender beta 8
